### ISSUE TITLE
Politics-coded speech in the feed — can we detect it, and should we? (large-scale research)

### ISSUE THREAD

**Issue by bmilojkovic** [goodgorithm/goodgorithm#228]
**Large-scale research effort — no committed direction, and the core premise below is itself one of the open questions.** This is a standalone investigation, not a scheduled feature. It is expected to run extensive experiments: training new classical models, evaluating on the [#177](https://github.com/goodgorithm/goodgorithm/issues/177) corpus and external datasets, not only ad-hoc DB queries. Distinct from [#192](https://github.com/goodgorithm/goodgorithm/issues/192) (vacuous / sarcasm / persuasion tail of surface-valence scoring) — that one is about hollowness; this one is about civic/political content as a category, whether or not it's "positive."

## The premise to validate (not assume)

Working hypothesis: **a positivity feed probably wants little or no politics-coded speech in it — most of the time.** The pitch is countering the negativity/outrage bias of mainstream algorithms, and political content on public social media skews heavily toward advocacy, grievance, dunking, and mobilization even when it scores positive on sentiment. So the intuition is that political-topic content is usually off-mission for this feed regardless of tone.

But this is an intuition, and the research should test it before any mechanism is built:

- **How much political-topic content actually reaches the ranked feed today?** (An early [#192](https://github.com/goodgorithm/goodgorithm/issues/192) probe found only ~26 high-sentiment ranked posts with political named entities and 0 in the top 300 — but that used a narrow entity list and one window. Needs a real measurement.)
- **What fraction of it is genuinely uplifting** vs. advocacy/outrage/mockery? (e.g. "my daughter registered to vote for the first time 🥹", a community organizer winning a local race, a naturalization-ceremony photo — these are civic *and* warm.)
- If the honest answer is "political content that reaches this feed is ~always low-value," a broad demote is defensible. If a meaningful slice is genuine civic joy, the false-positive cost of a blunt topic filter may not be worth it, and the answer is a documented limitation + manual moderation, like the marketing-domain gap.
- **Is this a product/mission decision that should be made explicitly first?** Whether — and how aggressively — the feed should exclude or demote civic content touches the Mission page directly. The research should surface the data to inform that call, not presume it.

## What "politics-coded speech" even means — definition work is part of the effort

These are different problems with very different tractability:

| Layer | Example | Classical feasibility |
|---|---|---|
| **Topic** — is this *about* politics/government/elections | "The Senate voted 51-49 on the appropriations bill" | Good. TF-IDF+LR, lexicon methods work. |
| **Stance / advocacy** — is this arguing a position | "Call your rep and tell them to vote no" | Hard. Needs framed supervision or transformers. |
| **Framing** — which moral/issue frame | economic vs. security vs. fairness framing of immigration | Hard. Media Frames Corpus territory. |
| **Outrage / mockery / grievance** — negative civic affect in positive-coded words | "😂😂 MAGA in a nutshell. Hilarious." | Very hard. Sarcasm-detection-adjacent; classical methods ~useless. |

The research should decide **which layer(s) are the actual target** (likely: topic, plus some notion of "charged" vs. "civic-neutral"), and **where the line sits** against genuine civic warmth.

## Candidate detection approaches (to be experimented, not pre-chosen)

Ordered roughly cheapest → most involved. All must respect **no LLM in the algorithm** (see below).

1. **Seed-lexicon → embedding-neighborhood expansion** (Empath, Fast et al. 2016; SocialSent / SentProp, Hamilton et al. 2016). Small hand-seed of unambiguous civic/political terms → nearest neighbors in `glove-twitter-100` (already in the stack, frozen, auditable) → a soft "political-ness" lexicon → score a post by lexicon coverage or by cosine of its mean word-vector to the seed centroid. Cheap, few labels, fully inspectable. Known weaknesses: static-embedding polysemy ("party", "right", "state", "bill", "house", "ice", "crown", "green"), and `glove-twitter-100` is ~2014-era so recent political vocab ("MAGA", "Brexit", "Dobbs", current surnames) is stale or absent — re-fitting embeddings on the [#177](https://github.com/goodgorithm/goodgorithm/issues/177) corpus is an experiment here.
2. **Nearest-centroid / prototype classification in embedding space** — mean-of-word-vectors post representation, "political prototype" from seeds or labeled posts, cosine score. Textbook weak baseline; cheap; useful as a floor to beat.
3. **A dedicated classical classifier** — TF-IDF + one-vs-rest / binary logistic regression, ONNX-exported, R2-versioned. This is a **copy of the existing category-classifier pattern** (`category_model.py` + `training/category_classifier.ipynb` + `training/r2_release.py`'s `MODEL_REGISTRY` + the `release-category-classifier` skill — all generalized by a `prefix` arg, not category-specific). Could be a 5th category (`civic_political`) or a separate binary devalue signal. **The blocker is training data** (see below).
4. **Reuse signals already extracted** — `topicality.py` runs spaCy NER with `NORP` in its label set (nationalities / religious / **political groups**); that label is currently discarded before storage. Political PERSON/ORG/NORP entity density is a free existing feature if the typed pairs are kept.
5. **Contextual sentence embeddings** (sentence-transformers / a small fine-tuned encoder) would fix the polysemy problem and materially help the stance/framing layers — **but this is the gray zone under "no LLM in the algorithm."** The constraint text explicitly blesses "a small CNN over word embeddings"; a transformer encoder as a core algorithm input is a different thing. The research can prototype with these to establish an *upper bound* on what's achievable, then report whether the gap to classical methods is large enough to warrant a constraint conversation.

## Data & experiments in scope

- **DB measurement** — prevalence, rank distribution, tone breakdown of political-topic content in `processed_posts` / `raw_posts`, across multiple windows.
- **A hand-labelled evaluation set** — ~500–1000 production posts labelled for {non-political, civic-neutral, civic-warm, advocacy/charged, outrage/mockery}. New infra (no per-post golden-scoring harness exists today; [#76](https://github.com/goodgorithm/goodgorithm/issues/76) added labelled *datasets* to the sentiment notebook only). This is a prerequisite for 3 and for evaluating 1–2. Could be shared with [#192](https://github.com/goodgorithm/goodgorithm/issues/192)'s labelling need.
- **The [#177](https://github.com/goodgorithm/goodgorithm/issues/177) corpus** — a domain-matched plain-text archive of feed-eligible content, currently accumulated but unconsumed. Usable as a background LM / IDF source, for re-fitting embeddings, and for weak-labelling at scale.
- **External datasets** — `cardiffnlp/tweet_topic_multi` (already used by the category classifier; has political-adjacent labels), the Media Frames Corpus, SemEval stance-detection sets, moral-outrage corpora (Brady et al.) — for pre-training / transfer / establishing what the literature ceiling is.
- **Embedding experiments** — `glove-twitter-100` as-is vs. re-fit on #177 vs. a from-scratch small embedding on #177; seed-set sensitivity; polysemy failure analysis.

## Constraints

- **No LLM in the algorithm.** Static word embeddings and classical classifiers are in-bounds. Contextual transformer encoders are the gray zone — prototype for an upper bound, don't ship without an explicit constraint decision.
- **No engagement signals.**
- **Precision over recall**, high false-positive bar — the existing discipline for `suppressed_terms` / `suppressed_domains` / `post_shapes`. A demote (like `aggregator_demote` / `hashtag_bag`) is lower-stakes than an exclude and is the likely mechanism if one is built.
- Must not punish genuine civic warmth. A naturalization photo, first-time-voter joy, a local volunteer drive — these are on-mission.
- Auditability: whatever ships has to be explainable ("this post was down-weighted because …"), consistent with the "trust why a post got selected" pitch. Favours an inspectable lexicon / linear model over an opaque one.

## Deliverables (research phase)

1. A measurement of how much political-topic content reaches the ranked feed and its tone breakdown — enough to answer "is the premise true."
2. A reusable labelled evaluation set.
3. Experimental results: for each approach tried, precision/recall on the labelled set, false-positive analysis on genuine civic content, and cost.
4. A recommendation: a concrete mechanism (with which layer it targets and at what strength), OR "documented limitation + manual moderation," OR "escalate to a Mission-level product decision first" — with the data to support whichever.

## Non-goals for now

- Shipping any scoring change before the measurement + labelled set exist.
- The stance/framing/outrage layers as a build target unless an experiment shows a classical method actually works on them.
- Anything requiring a transformer encoder in the live algorithm without a separate, explicit constraint conversation.
- Re-litigating [#192](https://github.com/goodgorithm/goodgorithm/issues/192)'s vacuous/sarcasm tail — related, separate.

**Comment by bmilojkovic** [comment]
Broken into 6 sub-issues (linked above), sequenced as a research pipeline rather than one big effort:

```
#232 (measure)   ──┐
                    ├──> #233 (labelled set) ──> #234 (baselines) ──> #235 (trained classifier, contingent)
                    │                                                       │
                    └──────────────────────────────────> #236 (upper-bound prototype, optional/low-priority)

#237 (corpus-scale experiments) — filed now, explicitly deferred/blocked
```

- **#232** and **#233** can start in parallel; #233's sampling benefits from #232's snapshots but doesn't hard-block on them.
- **#234** only needs #233's labelled set. It covers candidate approaches 1/2/4 from this issue — seed-lexicon→embedding-neighborhood expansion, nearest-centroid, and reusing the NER pipeline's already-extracted (but currently untyped-at-storage) `NORP` entities.
- **#235** — the dedicated classical classifier (candidate 3) — is explicitly contingent: only pursue it if #234's cheaper baselines miss the precision bar. If #234 already clears it, #235 gets closed as not needed rather than building a heavier model nobody asked for.
- **#236** is exploratory and lowest priority: a contextual-embedding prototype purely to measure the gap-to-classical-methods on the stance/framing/outrage layers, never a candidate for the live algorithm itself.
- **#237** holds the corpus-based experiments this issue originally proposed (re-fitting embeddings, weak-labelling at scale) — but the #177 corpus (`corpus_export.py`) shipped only days ago, production-only, with no monthly shard yet. There isn't enough accumulated data for those experiments to mean anything right now, so #237 stays parked until the corpus has real volume (order of months, not days).

Each sub-issue restates the constraints it inherits from here (no LLM in the live algorithm, no engagement signals, precision-over-recall/high false-positive bar, must not punish genuine civic warmth) so it's self-contained if picked up independently.

A rough sanity probe done while scoping this (a broader keyword regex than the original #192 entity-list probe, one snapshot) found ~1.5% of the ranked window matching political-adjacent vocabulary at a slightly-above-baseline average rank — with the top hits showing exactly the tension this issue describes: genuine civic warmth next to partisan dunking next to plain false positives from generic-word matches ("campaign" as in TTRPG campaign). That's not a measurement, just further evidence that #232 needs a real methodology, not another one-off keyword pass.

This issue itself stays open as the epic tracker; the final "deliverables" section (measurement + labelled set + experimental results + a recommendation) gets written back here once #232–#236 land.

**Comment by bmilojkovic** [comment]
Correction to the sub-issue split above: #232 (measurement) originally only planned manual repeated live-DB snapshots to work around the 24h retention window. #237 (corpus experiments) is exactly the mechanism built to survive past that window, so it should feed #232's prevalence/tone measurement too, not just the embedding-refit/weak-labelling work #228 lists it for. Updated #232 and #237 to say so explicitly: once the #177 corpus has real volume, #232 switches its prevalence/tone tracking to it instead of hand-triggered snapshots — rank-distribution comparison stays live-DB-only either way, since the corpus deliberately drops `rank_score`/`sentiment_score`. Also fixed the internal `228-A`..`228-F` labels in each sub-issue body to their actual issue numbers (#232–#237) now that they're assigned.

**Comment by bmilojkovic** [comment]
## Progress update — #232 and #233 closed

**#232 (measurement)** — 8 daily live-DB snapshots (2026-09-11 → 2026-09-18) answer the premise question. Aggregate tone across all snapshots (n=1094 lexicon-matched sightings): `civic-negative` 54.0%, `civic-neutral` 19.2%, `civic-warm` 15.2%, `non-political` 11.6% (false positives). Political-topic content reaches the ranked feed at a real ~1.1–1.7% background rate, stays near-absent from the very top (top 50/100) but is a steady presence in the broader pool, and its average `rank_score` sits *above* baseline in 6 of 8 snapshots — sentiment scoring does not suppress it, and emphatic outrage/mockery reads as positive-scoring "enthusiasm" as hypothesized. `civic-warm` is a real, consistent ~15% slice, not a rounding error. Also found: NORP-entity NER reuse is not a usable standalone signal (dominated by nationality-adjective noise), and the data shows clear event-driven bursts (AfD-Germany, Sweden election, the Massie/Hegseth motion, Carney's EU speech) rather than a flat rate. Full detail: #232.

**#233 (labelled evaluation set)** — 1181 maintainer-confirmed posts across the four-category taxonomy (`non-political` 190 / `civic-neutral` 248 / `civic-warm` 174 / `civic-negative` 569), well past the ~500–1000 target and the ~100-per-minority-class floor. Committed as a reusable artifact in #263: `training/political_tone_eval.jsonl` + `training/political_tone_eval_labeling_guide.md` (taxonomy, worked examples, boundary-case notes). Independently corroborates #232's aggregate breakdown (48.2/21.0/14.7/16.1 vs. #232's 54.0/19.2/15.2/11.6) — good cross-validation between the two efforts' different sampling approaches. Full detail: #233.

**Net so far**: the working hypothesis holds directionally (political content skews charged/negative, isn't naturally suppressed by sentiment scoring) but with the caveat #228 flagged as the real risk — genuine civic warmth is a consistent, non-trivial slice (~15%), so a blunt topic-level exclude isn't free.

**Next**: #234 (baseline experiments — seed-lexicon/embedding-neighborhood, nearest-centroid, NER-reuse) can now run against #233's committed eval set. #235/#236 remain contingent/exploratory per the original sequencing. This issue stays open as the epic tracker until #234–#236 land and the final recommendation gets written back here.

**Comment by bmilojkovic** [comment]
## #234 update — a candidate mechanism, with a sequencing recommendation

#234's baseline experiments (see the issue for full methodology) found that, among the three cheap approaches tested, **nearest-centroid is the clear winner for catching political content** — at a mistake budget of flagging ~8% of non-political posts wrongly, it catches 52% of political content on the held-out check set, vs. 16% for a propagated political-word lexicon and effectively 0% for NORP-entity presence alone.

Note this was evaluated under an explicit reprioritization: since political content is only ~1% of the feed (#232) and `civic-warm` is ~15% of that slice, catching political content was weighted far above protecting `civic-warm` from being caught too — a deliberate departure from #228's original "must not punish genuine civic warmth" framing, made for this evaluation. Nearest-centroid does also flag the majority of `civic-warm` content at the same rate it catches everything else (it's topic-only, no tone awareness) — see #234 for the exact number.

**Recommendation on sequencing**: nearest-centroid is a real, cheap, working candidate mechanism at this point. But #235 (a dedicated trained classical classifier, same TF-IDF+LR pattern as the category classifier) should be attempted first, since it may substantially outperform this baseline the same way #234's own results showed a fitted method (centroid) beating an unsupervised one (lexicon) by a wide margin — the same gap could repeat one tier up. If #235 clears nearest-centroid's numbers by a wide margin, there's no reason to also build and maintain the simpler nearest-centroid mechanism — build #235's classifier instead and treat nearest-centroid as a superseded baseline, not a second thing to ship.

**Comment by bmilojkovic** [comment]
New addition to this epic: #264 (implementation, building on #234/#235's closed research).
